openssl/src/base64.rs

```diff
@@ -0,0 +1,110 @@
+//! Utilities for base64 coding
+//!
+//! See manual page of [`EVP_EncodeInit`] for more information on the specific base64 variant.
+//!
+//! [`EVP_EncodeInit`]: https://www.openssl.org/docs/man1.1.1/man3/EVP_EncodeInit.html
+use error::ErrorStack;
+use ffi;
+use libc::c_int;
+use std::convert::TryInto;
+use cvt_n;
+
+/// Encodes a given block of bytes to base64.
+///
+/// # Panics
+///
+/// Panics if the input length or computed output length
+/// overflow a signed C integer.
+pub fn encode_block(src: &[u8]) -> String {
+    let src_len = src.len().try_into().unwrap();
+    let len = encoded_len(src_len).unwrap();
+    let mut out = Vec::new();
+    out.reserve(len);
+
+    // SAFETY: `encoded_len` ensures space for 4 output characters
+    // for every 3 input bytes including padding and nul terminator.
+    // `EVP_EncodeBlock` will write only single byte ASCII characters.
+    // `EVP_EncodeBlock` will only write to not read from `out`.
+    unsafe {
+        let out_len = ffi::EVP_EncodeBlock(out.as_mut_ptr(), src.as_ptr(), src_len);
+        out.set_len(out_len.try_into().unwrap());
+        String::from_utf8_unchecked(out)
+    }
+}
+
+/// Decodes a given base64-encoded text to bytes.
+///
+/// # Panics
+///
+/// Panics if the input length or computed output length
+/// overflow a signed C integer.
+pub fn decode_block(src: &str) -> Result<Vec<u8>, ErrorStack> {
+    let src_len = src.len().try_into().unwrap();
+    let len = decoded_len(src_len).unwrap();
+    let mut out = Vec::new();
+    out.reserve(len);
+
+    // SAFETY: `decoded_len` ensures space for 3 output bytes
+    // for every 4 input characters including padding.
+    // `EVP_DecodeBlock` can write fewer bytes after stripping
+    // leading and trailing whitespace, but never more.
+    // `EVP_DecodeBlock` will only write to not read from `out`.
+    unsafe {
+        let out_len = cvt_n(ffi::EVP_DecodeBlock(out.as_mut_ptr(), src.as_ptr(), src_len))?;
+        out.set_len(out_len.try_into().unwrap());
+        Ok(out)
+    }
+}
+
+fn encoded_len(src_len: c_int) -> Option<usize> {
```

Comment:
Why does this take a c_int and return a usize?

Reply:
I need the return value to `reserve`/`resize` the `out` buffer, so I need an `usize`. But the computation must succeed using a `c_int` for the OpenSSL calls to return valid counts otherwise I am not convinced I can rely on those counts for `set_len`/`truncate`.

In an earlier version, I used `c_int -> Option<c_int>` and had `encoded_len(src_len).unwrap().try_into().unwrap()` on the outside which seemed cumbersome but indeed less surprising. I can certainly changed this back if you prefer or I can add comment explaining the odd types better?

Reply:
But I just noticed that the CI builds failed as the use Rust 1.33 which I guess is this libraries current minimum supported Rust version so I guess I have to remove the `TryInto` usage in any case?

Reply:
Pushed a version that builds using Rust 1.33 and added a comment explaining the odd types.

Would you prefer the `usize_to_c_int` and `c_int_to_usize` in the top-level module or replaced by plain asserts as used in the other modules?

Reply:
It seems best to mirror the rest of the codebase.

Reply:
Pushed a version that hopefully better mirrors the style of the rest of the code base.